qlib/kernel/task.rs

```diff
@@ -165,10 +160,51 @@ impl Drop for Task {
     }
 }
 
+
+// task wrapper on shared memory
+#[repr(C)]
+pub struct TaskWrapper {
+    pub ready: AtomicU64,  // 0x0
+    // job queue id
+    pub queueId: AtomicUsize,
+    pub taskAddr: u64,
+}
+
+
+impl TaskWrapper {
+    pub fn New(task_addr: u64) -> Self {
+        return Self {
+            ready: AtomicU64::new(1),
+            queueId: AtomicUsize::new(0),
+            taskAddr: task_addr,
+        };
+    }
+
+    pub fn Ready(&self) -> u64 {
+        return self.ready.load(Ordering::Acquire);
+    }
+
+    pub fn SetReady(&self, val: u64) {
+        return self.ready.store(val, Ordering::SeqCst);
+    }
+
+    pub fn QueueId(&self) -> usize {
+        return self.queueId.load(Ordering::Acquire);
+    }
+
+    pub fn SetQueueId(&self, queueId: usize) {
+        return self.queueId.store(queueId, Ordering::Release);
+    }
+}
+
+
 #[repr(C)]
 pub struct Task {
     pub context: Context,
+    // address of Task struct
     pub taskId: u64,
+    // address of TaskWrapper struct
+    pub taskWrapperId: u64,
```

Comment:
Doesn't the new `struct TaskId` already have both fields?

Reply:
Sometimes, process can only access the `Struc Task` (e.g., using get_current_task()).   So `struct Task` also needs to record `taskWrapperId` so that the process can find `taskwrapper`